input/listen.go

```diff
@@ -8,126 +8,161 @@ import (
+
+		consume(addr)
+
+		select {
+		case <-l.shutdown:
+			return
+		default:
+			for {
+				log.Notice("reopening %v/%s", addr, proto)
+				err := reconnect(addr)
+				if err == nil {
+					backoffCounter.Reset()
+					break
+				}
+
+				backoffDuration := backoffCounter.Duration()
+				log.Error("error listening on %v/%s, retrying after %v: %s", addr, proto, backoffDuration, err)
+				select {
+				case <-l.shutdown:
+					log.Info("shutting down %v/%s, closing socket", addr, proto)
+					return
+				case <-time.After(backoffDuration):
+					// retry to connect
+				}
 			}
-
-			// handle the connection
-			log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
-			go acceptTcpConn(c, handler)
 		}
+	}
+}
 
-		for {
-			log.Notice("reopening %v/tcp", addr)
+func (l *Listener) listenTcp(addr string) error {
+	laddr, err := net.ResolveTCPAddr("tcp", addr)
+	if err != nil {
+		return err
+	}
+	l.tcpList, err = net.ListenTCP("tcp", laddr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-			l, err = listenTcp(addr)
-			if err == nil {
-				backoffCounter.Reset()
-				break
+func (l *Listener) acceptTcp(addr string) {
+	for {
+		c, err := l.tcpList.AcceptTCP()
+		if err != nil {
+			select {
+			// socket has been closed due to shut down, log and return
+			case <-l.shutdown:
+				return
+			default:
+				log.Error("error accepting on %v/tcp, closing connection: %s", addr, err)
+				l.tcpList.Close()
+				return
 			}
-
-			backoffDuration := backoffCounter.Duration()
-
-			log.Error("error listening on %v/tcp, retrying after %v: %s", addr, backoffDuration, err)
-			time.Sleep(backoffDuration)
 		}
+
+		// handle the connection
+		log.Debug("listen.go: tcp connection from %v", c.RemoteAddr())
+		l.wg.Add(1)
+		go l.acceptTcpConn(c)
 	}
 }
 
-func acceptTcpConn(c net.Conn, handler Handler) {
-	defer c.Close()
-	handler.Handle(c)
+func (l *Listener) acceptTcpConn(c net.Conn) {
+	defer l.wg.Done()
+
+	go func() {
+		<-l.shutdown
+		c.Close()
+	}()
+
+	l.handler.Handle(c)
 }
 
-func listenUdp(addr string) (*net.UDPConn, error) {
+func (l *Listener) listenUdp(addr string) error {
 	udp_addr, err := net.ResolveUDPAddr("udp", addr)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	l.udpConn, err = net.ListenUDP("udp", udp_addr)
+	if err != nil {
+		return err
 	}
-	return net.ListenUDP("udp", udp_addr)
+	return nil
 }
 
-func acceptUdp(addr string, udp_conn *net.UDPConn, handler Handler) {
-	var err error
-	l := udp_conn
+func (l *Listener) consumeUdp(addr string) {
 	buffer := make([]byte, 65535)
-	backoffCounter := &backoff.Backoff{
-		Min: 500 * time.Millisecond,
-		Max: time.Minute,
-	}
-
 	for {
-		log.Notice("listening on %v/udp", addr)
-
-		for {
-			// read a packet into buffer
-			b, src, err := l.ReadFrom(buffer)
-			if err != nil {
+		// read a packet into buffer
+		b, src, err := l.udpConn.ReadFrom(buffer)
+		if err != nil {
+			select {
+			// socket has been closed due to shut down, log and return
```

Comment:
this comment is incorrect. and if it were correct it would be redundent, so just remove it.